File: src/main/java/com/rtomyj/yugiohAPI/controller/stats/StatsController.java
package com.rtomyj.yugiohAPI.controller.stats;


import com.google.inject.internal.util.Strings;
import com.rtomyj.yugiohAPI.controller.YgoApiBaseController;
import com.rtomyj.yugiohAPI.dao.database.Dao;
import com.rtomyj.yugiohAPI.helper.constants.SwaggerConstants;
import com.rtomyj.yugiohAPI.model.Stats.DatabaseStats;
import com.rtomyj.yugiohAPI.model.Stats.MonsterType;
import io.swagger.annotations.Api;
import io.swagger.annotations.ApiOperation;
import io.swagger.annotations.ApiParam;
import io.swagger.annotations.ApiResponse;
import io.swagger.annotations.ApiResponses;
import lombok.NonNull;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.CrossOrigin;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import javax.servlet.http.HttpServletRequest;


/**
 *
 */
@RestController
@RequestMapping(path="/stats")
@CrossOrigin(origins="*")
@Slf4j
@Api(tags = SwaggerConstants.TAG_STATS_NAME)
public class StatsController extends YgoApiBaseController
{

    private static final String END_POINT = BASE_ENDPOINT + "/stats";

    private final Dao dao;


    @Autowired
    public StatsController(final HttpServletRequest request, @Qualifier("jdbc") final Dao dao)
    {

        this.request = request;
        this.dao = dao;

    }


    @ApiOperation(value = "Retrieve sum of all unique monster types for a given color of a card."
            , response = MonsterType.class
            , responseContainer = "Object")
    @ApiResponses(value = {
            @ApiResponse(code = 200, message = SwaggerConstants.http200)
            , @ApiResponse(code = 400, message = SwaggerConstants.http400)
            , @ApiResponse(code = 404, message = SwaggerConstants.http404)
    })
    @GetMapping("/card/monster_type/{cardColor}")
    public ResponseEntity<MonsterType> getMonsterTypeByColor(
            @ApiParam(
              value = SwaggerConstants.CARD_COLOR_DESCRIPTION
                    , defaultValue = "fusion"
            ) @NonNull @PathVariable("cardColor") final String cardColor
    )
    {

        return ResponseEntity.ok(dao.getMonsterTypeStats(Strings.capitalize(cardColor)));

    }


    @GetMapping()
    @ApiOperation(value = "Retrieve overview of the data currently in Database."
            , response = DatabaseStats.class
            , responseContainer = "Object")
    @ApiResponses(value = {
            @ApiResponse(code = 200, message = SwaggerConstants.http200)
            , @ApiResponse(code = 400, message = SwaggerConstants.http400)
            , @ApiResponse(code = 404, message = SwaggerConstants.http404)
    })
    public ResponseEntity<DatabaseStats> getDatabaseStats()
    {

        // TODO: add loggers
        log.info("Stat base endpoint hit");
        return ResponseEntity.ok(dao.getDatabaseStats());

    }

}

File: src/main/kotlin/com/rtomyj/skc/stats/StatsController.kt
package com.rtomyj.skc.stats

import com.rtomyj.skc.config.ReactiveMDC
import com.rtomyj.skc.config.SwaggerConfig
import com.rtomyj.skc.model.DatabaseStats
import com.rtomyj.skc.model.MonsterTypeStats
import com.rtomyj.skc.util.constant.SwaggerConstants
import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.Parameter
import io.swagger.v3.oas.annotations.enums.ParameterIn
import io.swagger.v3.oas.annotations.media.ExampleObject
import io.swagger.v3.oas.annotations.responses.ApiResponse
import io.swagger.v3.oas.annotations.tags.Tag
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import reactor.core.publisher.Mono

@RestController
@RequestMapping(path = ["/stats"], produces = ["application/json; charset=UTF-8"])
@Tag(name = SwaggerConstants.TAG_STATS_NAME)
class StatsController @Autowired constructor(private val statsService: StatsService) {
  companion object {
    private val log = LoggerFactory.getLogger(this::class.java.name)
  }

  @Operation(summary = "Retrieve sum of all unique monster types for a given color of a card.")
  @GetMapping("/card/monster_type/{cardColor}")
  @ApiResponse(responseCode = "200", description = SwaggerConstants.HTTP_200_SWAGGER_MESSAGE)
  @ApiResponse(responseCode = "400", ref = "badRequest")
  @ApiResponse(responseCode = "404", ref = "notFound")
  @ApiResponse(responseCode = "422", ref = "unprocessableEntity")
  @ApiResponse(responseCode = "500", ref = "internalServerError")
  fun monsterTypeStats(
    @Parameter(description = SwaggerConfig.CARD_COLOR_DESCRIPTION,
      `in` = ParameterIn.PATH,
      required = true,
      examples = [ExampleObject(ref = "fusion", name = "fusion"), ExampleObject(ref = "effect", name = "effect"),
        ExampleObject(ref = "synchro", name = "synchro")])
    @PathVariable("cardColor") cardColor: String
  ): Mono<MonsterTypeStats> = ReactiveMDC.deferMDC(
    Mono
        .fromCallable { statsService.getMonsterTypeStats(cardColor) }
        .doOnSubscribe {
          log.info("Retrieving monster types for cards with color: {}", cardColor)
        })

  @Operation(summary = "Retrieve overview of the data currently in Database.")
  @ApiResponse(responseCode = "200", description = SwaggerConstants.HTTP_200_SWAGGER_MESSAGE)
  @ApiResponse(responseCode = "400", ref = "badRequest")
  @ApiResponse(responseCode = "404", ref = "notFound")
  @GetMapping
  fun databaseStats(): Mono<DatabaseStats> = ReactiveMDC.deferMDC(
    Mono
        .fromCallable { statsService.databaseStats() }
        .doOnSuccess {
          log.info("Successfully retrieved database stats: {}", it.toString())
        }
        .doOnSubscribe { log.info("Retrieving high level overview of info stored in DB.") })
}